### DISCUSSION TITLE
Action Bundles

### DISCUSSION THREAD

**Discussion by with-heart** [statelyai/xstate#3554] [Ideas]
Hi yall! 👋 

*This discussion stems from a conversation with @Andarist regarding https://github.com/statelyai/xstate-tools/pull/207*

One major frustration I've faced when using `xstate` is the lack of type-safety when calling action functions from within `pure`. 

`pure` takes a function (`getActions`) that returns an array of actions to execute. Unfortunately, there's no way to inject the generic types `TContext`, `TEvent`, etc. that `pure` receives in the returned array, which means action function calls inside `pure` use their default generic types:

```ts
{
  actions: {
    // pure<TContext, TEvent> (or an event type derived from `eventsCausingActions` 
    // if using typegen)
    doSomething: pure(() => [
      // assign<unknown, EventObject>
      assign(),
      // send<unknown, EventObject>
      send(),
    ]);
  }
}
```

Since this makes `pure` difficult to work with, I'd like to propose an alternative that I've been thinking of as "action bundles". 

Instead of calling action functions within `pure`, we define the actions individually and then "bundle" them together by referring to them from another action:

```ts
{
  actions: {
    doSomething: ['assignThing', 'sendSomething'],
    // assign<TContext, TEvent>
    assignThing: assign(),
    // send<TContext, TEvent>
    sendSomething: send(),
  }
}
```

Much like the previous example using `pure`, executing `doSomething` will then execute `assignThing` and `sendSomething`. We get the same behavior but we retain type-safe awesomeness!

This also makes it much easier to typegen. https://github.com/statelyai/xstate-tools/pull/207 is an implementation for detecting array `type` strings returned in `pure` arrays, but since `getActions` is a function and a function return can take many shapes, it'd be very difficult for us to analyze all of the possible AST structures.

If actions can be defined as simple arrays of action `type` strings, this becomes relatively easy.

---

If you're familiar with `xstate`, you might be thinking to yourself "but @with-heart, you can already define actions as arrays of `type` strings using `entry`/`exit`/`actions` in the machine config, and those are all type-safe". And you'd be correct!

The unfortunate thing about that method is that sometimes you end up in situations where you have actions that should always be paired (executing one before or after the other, etc.), so you're stuck with the mental overhead of needing to remember those types of details. It works, but with enough actions, it's pretty frustrating to maintain.

**Comment by Andarist** [comment]
I think that this wouldn't bloat XState's code by a lot so perhaps it could be added to the core. We'd have to ensure that this would work in a recursive manner (one bundle could reference another bundle etc). I rarely needed something like this on my own though and I can imagine that naming such bundles could be a problem.

This will also have some impact on the Studio in the future as it would have to understand this pattern, edit it, etc. It's worth waiting for @davidkpiano's take on that.

**Reply by davidkpiano** [reply]
I think it's a natural fit; I've wanted this feature as well. It makes sense that certain actions can be grouped together; after all, if states can be hierarchical, why not extend that feature to actions? 💡 

I don't foresee this being a problem in the studio. An action object is just an object, and in the future, editing an action would have a generated form (based on a JSON Schema) for each action, whether it's user-defined or built-in. For these "action groups" specifically, I'd imagine that `foo: ['bar', 'baz']` would be sugar for:

```js
// pseudocode
// foo: ['bar', 'baz']

foo: {
  type: 'xstate.actionGroup',
  actions: [
    { type: 'bar' },
    { type: 'baz' }
  ]
}
```

**Reply by with-heart** [reply]
Awesome! Love to hear that!

I started playing around with implementations for `xstate` and typegen, should have PRs tomorrow 😁
